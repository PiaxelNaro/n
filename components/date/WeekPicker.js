import React, { Component } from 'react';
import { DayPicker } from 'react-dates';
import './style.scss';

export default class WeekPicker extends Component {
    props: {
        onDateSelected: () => any,
        currentStartDate?: moment$Moment,
        currentEndDate?: moment$Moment,
    };

    state: {
        hoveredDay: ?moment$Moment,
    };

    onDayClick: Function;
    isDayHovered: Function;
    isDaySelected: Function;
    onDayMouseEnter: Function;
    onDayMouseLeave: Function;

    constructor() {
        super();

        this.state = {
            hoveredDay: null,
        };

        this.onDayClick = this.onDayClick.bind(this);
        this.isDayHovered = this.isDayHovered.bind(this);
        this.isDaySelected = this.isDaySelected.bind(this);
        this.onDayMouseEnter = this.onDayMouseEnter.bind(this);
        this.onDayMouseLeave = this.onDayMouseLeave.bind(this);
    }

    isDayHovered(day: moment$Moment) {
        const hoverdDay = this.state.hoveredDay;

        return Boolean(hoverdDay && day.isSame(hoverdDay, 'week'));
    }

    isDaySelected(day: moment$Moment) {
        const { currentStartDate, currentEndDate } = this.props;
        if (currentStartDate && currentEndDate) {
            return (
                day.isSameOrAfter(currentStartDate) && day.isSameOrBefore(currentEndDate)
            );
        }

        return false;
    }

    onDayMouseEnter(day: moment$Moment) {
        this.setState({ hoveredDay: day.clone().utc() });
    }

    onDayMouseLeave() {
        this.setState({ hoveredDay: null });
    }

    onDayClick(day: moment$Moment) {
        const startDate = day.clone().utc().startOf('week');
        const endDate = day.clone().utc().endOf('week');
        this.props.onDateSelected(startDate, endDate);
    }

    render() {
        return (
            <DayPicker
                numberOfMonths={3}
                daySize={28}
                modifiers={{
                    hovered: this.isDayHovered,
                    selected: this.isDaySelected,
                }}
                onDayMouseEnter={this.onDayMouseEnter}
                onDayMouseLeave={this.onDayMouseLeave}
                onDayClick={this.onDayClick}
            />
        );
    }
}
