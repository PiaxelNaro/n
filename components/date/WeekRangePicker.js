// @flow
import React from 'react';
import { DayPicker } from 'react-dates';

const orderAndRoundDates = (date1: moment$Moment, date2: moment$Moment) => {
    let startDate = date1;
    let endDate = date2;
    if (date2.isBefore(date1)) {
        startDate = date2;
        endDate = date1;
    }

    return {
        startDate: startDate.startOf('week'),
        endDate: endDate.endOf('week'),
    };
};

class WeekRangePicker extends React.Component {
    props: {
        onDateSelected: (moment$Moment, moment$Moment) => any,
        currentStartDate?: moment$Moment,
        currentEndDate?: moment$Moment,
    };

    state: {
        clickedDate: ?moment$Moment,
        hoveredDay: ?moment$Moment,
    };

    onDayClick: Function;
    isDayHovered: Function;
    isDaySelected: Function;
    onDayMouseEnter: Function;
    onDayMouseLeave: Function;

    constructor() {
        super();

        this.state = {
            clickedDate: null,
            hoveredDay: null,
        };

        this.onDayClick = this.onDayClick.bind(this);
        this.isDayHovered = this.isDayHovered.bind(this);
        this.isDaySelected = this.isDaySelected.bind(this);
        this.onDayMouseEnter = this.onDayMouseEnter.bind(this);
        this.onDayMouseLeave = this.onDayMouseLeave.bind(this);
    }

    isDayHovered(day: moment$Moment) {
        const date1 = this.state.clickedDate;
        const date2 = this.state.hoveredDay;

        if (date1 && date2) {
            const { startDate, endDate } = orderAndRoundDates(date1, date2);
            return day.isBetween(startDate, endDate);
        }

        return Boolean(
            (date1 && day.isSame(date1, 'week')) || (date2 && day.isSame(date2, 'week'))
        );
    }

    isDaySelected(day: moment$Moment) {
        // We don't display selected days when the user is in the progress of selecting a new range
        if (this.state.clickedDate) {
            return false;
        }

        const { currentStartDate, currentEndDate } = this.props;
        if (currentStartDate && currentEndDate) {
            return (
                day.isSameOrAfter(currentStartDate) && day.isSameOrBefore(currentEndDate)
            );
        }

        return false;
    }

    onDayMouseEnter(day: moment$Moment) {
        this.setState({ hoveredDay: day.clone().utc() });
    }

    onDayMouseLeave() {
        this.setState({ hoveredDay: null });
    }

    onDayClick(day: moment$Moment) {
        const clickedDate = day.clone().utc();
        if (!this.state.clickedDate) {
            this.setState({ clickedDate });
            return;
        }

        const { startDate, endDate } = orderAndRoundDates(
            this.state.clickedDate,
            clickedDate
        );
        this.props.onDateSelected(startDate, endDate);
        this.setState({ hoveredDay: null, clickedDate: null });
    }

    render() {
        return (
            <DayPicker
                numberOfMonths={3}
                daySize={28}
                modifiers={{
                    hovered: this.isDayHovered,
                    selected: this.isDaySelected,
                }}
                onDayMouseEnter={this.onDayMouseEnter}
                onDayMouseLeave={this.onDayMouseLeave}
                onDayClick={this.onDayClick}
            />
        );
    }
}

export default WeekRangePicker;
