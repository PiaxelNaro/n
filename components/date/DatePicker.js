// @flow
import React from 'react';
import Head from 'next/head';
import CalendarIcon from 'react-icons/lib/fa/calendar';
import WeekRangePicker from './WeekRangePicker';
import WeekPicker from './WeekPicker';
import Overlay from '../Overlay';
import DropdownButton from '../Button/Dropdown';
import formatDateRange from './util/formatDateRange';
import cssStyle from './style.scss';
import { translate, type I18nType } from '../../modules/i18n';

const DatePickerSwitch = ({
    granularity,
    onDateSelected,
    startDate,
    endDate,
}: {
    startDate: moment$Moment,
    endDate: moment$Moment,
    onDateSelected: (moment$Moment, moment$Moment) => any,
    granularity: string,
}) => {
    let picker;
    switch (granularity) {
        case 'WeekRange':
            picker = (
                <WeekRangePicker
                    onDateSelected={onDateSelected}
                    currentStartDate={startDate}
                    currentEndDate={endDate}
                />
            );
            break;
        case 'Week':
            picker = (
                <WeekPicker
                    onDateSelected={onDateSelected}
                    currentStartDate={startDate}
                    currentEndDate={endDate}
                />
            );
            break;
        default:
            throw new Error(`Invalid granularity "${granularity}" passed to DatePicker`);
    }

    return picker;
};

class DatePicker extends React.Component {
    props: {
        startDate: moment$Moment,
        endDate: moment$Moment,
        onDateSelected: (moment$Moment, moment$Moment) => any,
        granularity: string,
        i18n: I18nType,
    };

    overlay: { closeOverlay: () => void };
    onDateSelected: Function;

    constructor() {
        super();

        this.onDateSelected = this.onDateSelected.bind(this);
    }

    onDateSelected(startDate: moment$Moment, endDate: moment$Moment) {
        if (this.overlay) {
            this.overlay.closeOverlay();
        }
        this.props.onDateSelected(startDate, endDate);
    }

    render() {
        const { startDate, endDate, granularity } = this.props;
        const { gettext } = this.props.i18n;

        return (
            <Overlay.Trigger
                trigger="click"
                ref={ref => (this.overlay = ref)}
                getOverlay={() =>
                    <Overlay>
                        <DatePickerSwitch
                            granularity={granularity}
                            onDateSelected={this.onDateSelected}
                            startDate={startDate}
                            endDate={endDate}
                        />
                    </Overlay>}
            >
                <Head>
                    <style data-source="aa-react-date-picker-css" type="text/css">
                        {cssStyle}
                    </style>
                </Head>
                <DropdownButton
                    label={gettext('Date Range')}
                    labelIcon={<CalendarIcon size={11} />}
                    value={formatDateRange(startDate, endDate)}
                    width="auto"
                />
            </Overlay.Trigger>
        );
    }
}

export { DatePicker as _DatePicker, DatePickerSwitch };
export default translate(DatePicker);
