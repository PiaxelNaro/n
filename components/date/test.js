// @flow
import React from 'react';
import moment from 'moment';
import { shallow, mount } from 'enzyme';
import range from 'lodash/range';
import { _DatePicker as DatePicker, DatePickerSwitch } from './DatePicker';
import WeekRangePicker from './WeekRangePicker';
import WeekPicker from './WeekPicker';
import mockI18n from '../../modules/i18n/mockI18n';

describe('<DatePicker />', () => {
    it('handles date range selection', () => {
        const spy = jest.fn();
        const datePicker = shallow(
            <DatePicker
                startDate={moment.utc([2014, 4, 5])}
                endDate={moment.utc([2016, 4, 5])}
                onDateSelected={spy}
                granularity="WeekRange"
                i18n={mockI18n}
            />
        );

        const overlay = shallow(datePicker.find('OverlayTrigger').prop('getOverlay')());
        overlay.find('DatePickerSwitch').prop('onDateSelected')();
        expect(spy).toBeCalled();
    });

    it('closes overlay once date is selected', () => {
        const spy = jest.fn();
        const startDate = moment();
        const endDate = moment();
        const datePicker = mount(
            <DatePicker
                startDate={moment.utc([2014, 4, 5])}
                endDate={moment.utc([2016, 4, 5])}
                onDateSelected={spy}
                granularity="WeekRange"
                i18n={mockI18n}
            />
        );

        const inst: any = datePicker.instance();
        inst.overlay = { closeOverlay: jest.fn() };

        const overlay = shallow(datePicker.find('OverlayTrigger').prop('getOverlay')());
        overlay.find('DatePickerSwitch').prop('onDateSelected')(startDate, endDate);
        expect(spy).toBeCalledWith(startDate, endDate);
        expect(inst.overlay.closeOverlay).toHaveBeenCalled();
    });

    it("select date even if overlay isn't present", () => {
        const spy = jest.fn();
        const startDate = moment();
        const endDate = moment();
        const datePicker = shallow(
            <DatePicker
                startDate={moment.utc([2014, 4, 5])}
                endDate={moment.utc([2016, 4, 5])}
                onDateSelected={spy}
                granularity="WeekRange"
                i18n={mockI18n}
            />
        );

        const inst: any = datePicker.instance();
        inst.onDateSelected(startDate, endDate);
        expect(spy).toBeCalledWith(startDate, endDate);
    });
});

describe('<DatePickerSwitch />', () => {
    it('render WeekRangePicker', () => {
        expect(
            shallow(
                <DatePickerSwitch
                    startDate={moment.utc([2014, 4, 5])}
                    endDate={moment.utc([2016, 4, 5])}
                    onDateSelected={() => {}}
                    granularity="WeekRange"
                />
            )
        ).toMatchSnapshot();
    });

    it('render WeekPicker', () => {
        expect(
            shallow(
                <DatePickerSwitch
                    startDate={moment.utc([2017, 3, 2])}
                    endDate={moment.utc([2017, 3, 8])}
                    onDateSelected={() => {}}
                    granularity="Week"
                />
            )
        ).toMatchSnapshot();
    });

    it('does not allow unsupported granularity type', () => {
        expect(() => {
            shallow(
                <DatePickerSwitch
                    startDate={moment.utc([2014, 4, 5])}
                    endDate={moment.utc([2016, 4, 5])}
                    onDateSelected={() => {}}
                    granularity="Does-NOT-exists"
                />
            );
        }).toThrow();
    });
});

describe('<WeekRangePicker />', () => {
    it('marks all days of hovered week as hovered', () => {
        const tuesday = moment.utc([2017, 3, 18]);
        const picker = shallow(<WeekRangePicker onDateSelected={() => {}} />);
        const controlledPicker = picker.find('DayPicker');

        const hovered = controlledPicker.prop('modifiers').hovered;
        const selected = controlledPicker.prop('modifiers').selected;

        // No hover, no pre-selected date. No dates are selected or hovered
        expect(hovered(tuesday)).toEqual(false);
        expect(selected(tuesday)).toEqual(false);

        controlledPicker.prop('onDayMouseEnter')(tuesday);
        range(16, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(hovered(date)).toEqual(true);
        });
        expect(hovered(moment.utc([2017, 3, 15]))).toEqual(false);
        expect(hovered(moment.utc([2017, 3, 23]))).toEqual(false);
        expect(hovered(tuesday.clone().add(1, 'month'))).toEqual(false);
    });

    it('marks all days of clicked week as hovered', () => {
        const tuesday = moment.utc([2017, 3, 18]);
        const picker = shallow(<WeekRangePicker onDateSelected={() => {}} />);
        const controlledPicker = picker.find('DayPicker');

        controlledPicker.prop('onDayClick')(tuesday);
        const hovered = controlledPicker.prop('modifiers').hovered;

        range(16, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(hovered(date)).toEqual(true);
        });
        expect(hovered(moment.utc([2017, 3, 15]))).toEqual(false);
        expect(hovered(moment.utc([2017, 3, 23]))).toEqual(false);
        expect(hovered(tuesday.clone().add(1, 'month'))).toEqual(false);
    });

    it('marks all days in between two dates as hovered', () => {
        const thursday = moment.utc([2017, 3, 6]);
        const tuesday = moment.utc([2017, 3, 18]);
        const picker = shallow(<WeekRangePicker onDateSelected={() => {}} />);
        const controlledPicker = picker.find('DayPicker');

        controlledPicker.prop('onDayClick')(thursday);
        controlledPicker.prop('onDayMouseEnter')(tuesday);
        const hovered = controlledPicker.prop('modifiers').hovered;

        range(3, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(hovered(date)).toEqual(true);
        });
        expect(hovered(moment.utc([2017, 3, 1]))).toEqual(false);
        expect(hovered(moment.utc([2017, 3, 23]))).toEqual(false);
        expect(hovered(tuesday.clone().add(1, 'month'))).toEqual(false);

        controlledPicker.prop('onDayMouseLeave')();
    });

    it('marks all days between selected range as selected until user click a new date', () => {
        const monday = moment.utc([2017, 3, 1]);
        const sunday = moment.utc([2017, 3, 23]);
        const picker = shallow(
            <WeekRangePicker
                onDateSelected={() => {}}
                currentStartDate={monday}
                currentEndDate={sunday}
            />
        );
        const controlledPicker = picker.find('DayPicker');
        const selected = controlledPicker.prop('modifiers').selected;

        range(1, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(selected(date)).toEqual(true);
        });
        expect(selected(sunday.clone().add(1, 'month'))).toEqual(false);

        // Once we click a new date, then no date is selected
        controlledPicker.prop('onDayClick')(moment.utc([2017, 4, 14]));
        expect(selected(monday)).toEqual(false);
        expect(selected(sunday)).toEqual(false);
    });

    it('allows to select date range in reverse order', () => {
        const thursday = moment.utc([2017, 3, 6]);
        const tuesday = moment.utc([2017, 3, 18]);
        const spy = jest.fn();
        const picker = shallow(<WeekRangePicker onDateSelected={spy} />);
        const controlledPicker = picker.find('DayPicker');

        controlledPicker.prop('onDayClick')(tuesday);
        controlledPicker.prop('onDayClick')(thursday);
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy.mock.calls[0][0].format('l')).toEqual('4/2/2017');
        expect(spy.mock.calls[0][1].format('l')).toEqual('4/22/2017');
    });
});

describe('<WeekPicker />', () => {
    it('marks all days of hovered week as hovered', () => {
        const tuesday = moment.utc([2017, 3, 18]);
        const picker = shallow(<WeekPicker onDateSelected={() => {}} />);
        const controlledPicker = picker.find('DayPicker');

        const hovered = controlledPicker.prop('modifiers').hovered;
        const selected = controlledPicker.prop('modifiers').selected;

        // No hover, no pre-selected date. No dates are selected or hovered
        expect(hovered(tuesday)).toEqual(false);
        expect(selected(tuesday)).toEqual(false);

        controlledPicker.prop('onDayMouseEnter')(tuesday);
        range(16, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(hovered(date)).toEqual(true);
        });
        expect(hovered(moment.utc([2017, 3, 15]))).toEqual(false);
        expect(hovered(moment.utc([2017, 3, 23]))).toEqual(false);
        expect(hovered(tuesday.clone().add(1, 'month'))).toEqual(false);

        controlledPicker.prop('onDayMouseLeave')();

        range(16, 22).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(hovered(date)).toEqual(false);
        });
    });

    it('marks all days in the same week as selected', () => {
        const startDate = moment.utc([2017, 3, 2]);
        const endDate = moment.utc([2017, 3, 8]);
        const spy = jest.fn();
        const picker = shallow(
            <WeekPicker
                onDateSelected={spy}
                currentStartDate={startDate}
                currentEndDate={endDate}
            />
        );
        const controlledPicker = picker.find('DayPicker');
        const selected = controlledPicker.prop('modifiers').selected;
        range(2, 8).forEach(day => {
            const date = moment.utc([2017, 3, day]);
            expect(selected(date)).toEqual(true);
        });
        expect(selected(startDate.clone().add(1, 'month'))).toEqual(false);

        const newDate = moment.utc([2017, 3, 9]);
        controlledPicker.prop('onDayClick')(newDate);
        const newStartDate = newDate.clone().utc().startOf('week');
        const newEndDate = newDate.clone().utc().endOf('week');
        expect(spy).toBeCalledWith(newStartDate, newEndDate);
    });
});
