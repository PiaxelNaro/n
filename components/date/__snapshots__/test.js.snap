// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerSwitch /> render WeekPicker 1`] = `
<WeekPicker
  currentEndDate={"2017-04-08T00:00:00.000Z"}
  currentStartDate={"2017-04-02T00:00:00.000Z"}
  onDateSelected={[Function]}
/>
`;

exports[`<DatePickerSwitch /> render WeekRangePicker 1`] = `
<WeekRangePicker
  currentEndDate={"2016-05-05T00:00:00.000Z"}
  currentStartDate={"2014-05-05T00:00:00.000Z"}
  onDateSelected={[Function]}
/>
`;
